systemd-nspawn: Allow setting ambient capability set

myrkr opened the pull request:
The old code was only able to pass the value 0 for the inheritable
and ambient capability set when a non-root user was specified.

However, sometimes it is useful to run a program in its own container
with a user specification and some capabilities set. This is needed
when the capabilities cannot be provided by file capabilities (because
the file system is mounted with MS_NOSUID for additional security).

This commit introduces the option --ambient-capability and the config
file option AmbientCapability=. Both are used in a similar way to the
existing Capability= setting. It changes the inheritable and ambient
set (which is 0 by default). The code also checks that the settings
for the bounding set (as defined by Capability= and DropCapability=)
and the setting for the ambient set (as defined by AmbientCapability=)
are compatible. Otherwise, the operation would fail in any way.

Due to the current use of -1 to indicate no support for ambient
capability set the special value "all" cannot be supported.

Also, the setting of ambient capability is restricted to running a
single program in the container payload.

poettering commented on man/systemd-nspawn.xml:
```diff
@@ -1017,6 +1017,26 @@
+        supported for this setting.</para>
+
+        <para>All capabilities specified here must be in the set
+        allowed with the --capability and --drop-capability
```
enclose the two switches in `<option>` please. Also, suffix them with "=", as we usually do for options which accept parameters. i.e. `<option>--capability=</option>`.

poettering commented on man/systemd-nspawn.xml:
```diff
@@ -1017,6 +1017,26 @@
+        options. Otherwise, an error message will be shown.</para>
+
+        <para>This option cannot be combined with the boot mode of the
+        container.</para>
```
add `(as requested via <option>--boot</option>)`

poettering commented on man/systemd-nspawn.xml:
```diff
@@ -1017,6 +1017,26 @@
+        <para>If the special value of <literal>help</literal> is
+        passed, the program will print known capability names and
+        exit.</para></listitem>
+      </varlistentry>
```
hmm, so far we got away with not specifying in the explanations for --capability= and --capability-drop= which caps sets these optoins affect. But if there's now control for the ambient caps, we need to be clearer for the other option too i guess. hence please update the relevant explanations about this too. this can be brief, key is just to say they affect the caps bounding set (and thus also the other two).

poettering commented on src/nspawn/nspawn-settings.h:
```diff
@@ -156,6 +156,7 @@ typedef struct Settings {
         char *user;
         uint64_t capability;
+        uint64_t ambient_capability;
         uint64_t drop_capability;
```
`capability` and `drop_capability` both affect the same set of caps, while `ambient_capability` affects a different set. hence i'd swap the order here, and put ambient_capability last

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -1018,6 +1024,15 @@ static int parse_argv(int argc, char *argv[]) {
+                        r = parse_capability_spec(optarg, &m);
+                        if (r <= 0)
+                                return r;
+                        arg_caps_ambient = m;
```
hmm, i think this should OR things in. In almost all cases our command line switches extend each other if that makes any conceptual sense, and hence we should do the same here.

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -4098,6 +4125,13 @@ static int merge_settings(Settings *settings, const char *path) {
+                        log_warning("Ignoring AmbientCapability= setting, file %s is not trusted.", path);
+                } else {
+                        arg_caps_ambient |= ambient;
+                }
```
please follow coding style, no {} around single-line if blocks

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -4098,6 +4125,13 @@ static int merge_settings(Settings *settings, const char *path) {
                 }
+
+                ambient = settings->ambient_capability;
+                if (! arg_settings_trusted && ambient != 0) {
```
we'd normally not place a space after the `!`

poettering reviewed (changes requested):
looks good, just some minor comments

myrkr commented:
I have updated my branch to address your comments. I have also added a remark regarding the dependency of Capability=/DropCapability= and AmbientCapability= via the bounding set to the systemd.nspawn.xml file.

poettering commented on src/nspawn/nspawn.c:
```diff
@@ -378,6 +379,9 @@ static int help(void) {
+               "     --ambient-capability=CAP\n"
+               "                            Sets the specified capability for the started\n"
+               "                            process. Not useful if booting a machine.\n"
                "     --drop-capability=CAP  Drop the specified capability from the default set\n"
```
sorry for not noticing this earlier, but here too, please swap the latter two options, i.e. list `--capability=`/`--drop-capabaility=` first, and `--amvient-capability=` after that, since the first two operate on the same set of caps, and the other on another

poettering reviewed (changes requested):
one more minor fix, please.

poettering commented:
lgtm. thanks!
